### PR TITLE
Find stages (whether using markers or block scopes)

### PR DESCRIPTION
**Pull request by svanoort**
Downstream of https://github.com/jenkinsci/workflow-api-plugin/pull/10

*  Defines a ChunkFinder implementation that will mark chunk start/end for both kinds of stage implementations (but ignores labelled parallel branches)
* Dependencies needed to identify block-scoped stages (ex StepStart/EndNode) mean this can't live in workflow-API

Upstream of block-scoped stage support in stage view. 

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+public class StageChunkFinder implements ChunkFinder {
+
+    public boolean isStartInsideChunk() {
+        return true;
+    }
+
+    @Override
+    public boolean isChunkStart(@Nonnull FlowNode current, @CheckForNull FlowNode previous) {
+        LabelAction label = current.getAction(LabelAction.class);
+        return label != null && !(label instanceof ThreadNameAction);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: :ant: Seems too implementation-dependent. Better to check that

```java
current.getAction(ThreadNameAction.class) == null
```

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    public boolean isChunkEnd(@Nonnull FlowNode current, @CheckForNull FlowNode previous) {
+        // First a block-scoped stage
+        if (current instanceof StepEndNode && ((StepEndNode) current).getDescriptor() instanceof StageStep.DescriptorImpl) {
+            // We have to look for the labelaction because block-scoped stage creates two nested blocks
+            return ((StepEndNode) current).getStartNode().getAction(LabelAction.class) != null;
+        }
+        // Then a marker-scoped stage
+        if (previous != null) {
+            LabelAction label = previous.getAction(LabelAction.class);
+            return label != null && !(label instanceof ThreadNameAction);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: :ant: ditto

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTiming.java [hunk 3]
```diff
@@ -365,4 +371,64 @@ public static GenericStatus condenseStatus(@Nonnull Collection<GenericStatus> st
...
+        System.out.println("\t- actionClassName actionDisplayName");
+        System.out.println("------------------------------------------------------------------------------------------");
+        for (FlowNode node : sorted) {
+            StringBuilder formatted = new StringBuilder();
+            formatted.append('[').append(node.getId()).append(']');
+            formatted.append('{').append(StringUtils.join(node.getParentIds(), ',')).append('}');
+            if (showTiming) {
+                formatted.append('(');
+                if (node.getAction(TimingAction.class) != null) {
+                    formatted.append(TimingAction.getStartTime(node)-startTime);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTiming.java]: Consider using `TimingAction.getStartTime`.

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTiming.java [hunk 4]
```diff
@@ -365,4 +371,64 @@ public static GenericStatus condenseStatus(@Nonnull Collection<GenericStatus> st
         }
         return Collections.max(statuses);
     }
+
+    /** Helper, prints flow graph in some detail - now a common utility so others don't have to reinvent it */
+    public static void printNodes(@Nonnull FlowExecution exec, long startTime, boolean showTiming, boolean showActions) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StatusAndTiming.java]: Suggest using

```java
@Restricted(DoNotUse.class)
```

so it will not be offered to non-test code.

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+/**
+ * Tests the stage collection
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public class StageTest {
+
+    /** Assert that chunk flownode IDs match expected, use 0 or -1 ID for null flownode */
+    public static void assertChunkBoundary(FlowChunkWithContext chunk, int beforeId, int firstId, int lastId, int afterId) {
+        // First check the chunk boundaries, then the before/after
+        Assert.assertNotNull(chunk.getFirstNode());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: Most people prefer to static-import `Assert`.

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                "node {" +
+                "   stage ('Build') { " +
+                "     echo ('Building'); " +
+                "   } \n" +
+                "   stage ('Test') { " +
+                "     echo ('Testing'); " +
+                "   } \n" +
+                "   stage ('Deploy') { " +
+                (hudson.Functions.isWindows() ? "   bat ('rmdir /s/q targs || echo no such dir\\n mkdir targs && echo hello> targs\\\\hello.txt'); "
+                        : "   sh ('rm -rf targs && mkdir targs && echo hello > targs/hello.txt'); "
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: Is it really necessary to run external processes for this test?

Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: (`writeFile` would be shorter and more portable.)

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+        job.setDefinition(new CpsFlowDefinition("" +
+                "node {" +
+                "   stage ('Build'); " +
+                "   echo ('Building'); " +
+                "   stage ('Test'); " +
+                "   echo ('Testing'); " +
+                "   stage ('Deploy'); " +
+                (hudson.Functions.isWindows() ? "   bat ('rmdir /s/q targs || echo no such dir\\n mkdir targs && echo hello> targs\\\\hello.txt'); "
+                        : "   sh ('rm -rf targs && mkdir targs && echo hello > targs/hello.txt'); "
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: ditto

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+                ) +
+                "   archive(includes: 'targs/hello.txt'); " +
+                "   echo ('Deploying'); " +
+                "}"));
+        /**
+         * Node dump follows, format:
+         [ID]{parent,ids} flowNodeClassName stepDisplayName [st=startId if a block node]
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: You may want to think about creating a more pleasant assertion utility that lets you encode the expected structure in the test code (not a comment) and not have to patch up your test code every time ID numbers change. A bit more work up front but will probably pay off in legibility and ease of maintenance.

**Comment by jglick** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java (hunk 1 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: @jglick Are there plans to change the underlying implementation in the near future?

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+
+/**
+ * Tests the stage collection
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
```
Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: 🐜 Javadoc linter wil unlikely agree with this syntax

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+ */
+public class StageTest {
```
Comment (oleg-nenashev) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: Strange ordering of fields/methods in this class

**Comment by oleg-nenashev** [comment]
🐝 with minor comments

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java (hunk 6 above)
Comment (svanoort) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: True, I'll change that.

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java (hunk 8 above)
Comment (svanoort) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: Not sure what you mean?  I think I would need to see at least pseudo-syntax to follow what specifically you're describing here.

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java (hunk 1 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: Not particularly, but no reason to make your code unnecessarily fragile.

Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: It was intended to improve performance slightly, but... I think at this point a premature optimization since there won't be that many nodes with a LabelAction.   You've convinced me, making the change.

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java (hunk 8 above)
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: Maybe something like

```java
assertChunksAndStages(build,
    "Build:FlowStartNode," +
        "StepStartNode(node)+LogActionImpl+WorkspaceActionImpl)," +
        "StepStartNode(node)+BodyInvocationAction,…",
    "Test:…",
    "Deploy:…");
```

**Comment by jglick** [comment]
re:bee:

**Comment by oleg-nenashev** [comment]
re:bee:

File: src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java (hunk 9 above)
Comment (svanoort) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: It seems to generate the docs fine (though the link isn't correct, as noted). 

I probably should just strip out the author tags, since it's a single-author plugin so far.

Comment (svanoort) [inline src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java]: In order to not clutter this PR up with extraneous changes, that would be a separate PR though.
